### ISSUE TITLE
C.4 - exception on overload sets

### ISSUE THREAD

**Issue by viboes** [isocpp/CppCoreGuidelines#1264]
In http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c4-make-a-function-a-member-only-if-it-needs-direct-access-to-the-representation-of-a-class

There is a mention of this exception

Exception

An overload set may have some members that do not directly access private data:
```
class Foobar {
public:
    void foo(long x)    { /* manipulate private data */ }
    void foo(double x) { foo(std::lround(x)); }
    // ...
private:
    // ...
};
```

I'm not sure the rationale is complete as we could overload non-free functions as in 

```
class Foobar {
public:
    void foo(long x)    { /* manipulate private data */ }
    // ...
private:
    // ...
};


void foo(Foobar& that, long x)    { that.foo(x); }
void foo(Foobar& that, double x) { foo(that, std::lround(x)); }
```

I recognize that this could be cumbersome if we had a lot of member overloads, but there is an alternative that IMO it is worth mentioning when we want to compare and the number of overloaded member is small.

Do you believe it is worth proposing a PR for this issue?

**Comment by viboes** [comment]
These are all good questions.

I'm suggesting that we should remove the exception and also the enforcement to ignore those kind of functions.

However, we could consider the case when having  a lot of member functions that are overloaded that use the data, that adding an overload to this set that don't use the data is an acceptable solution as adding the free overloads could be cumbersome. This can be seen as an exception also.

In summary, I believe that what we could do is to refine the current exception, and make it more limited in scope. 

The enforcement should change in accordance with the exception.

**Comment by hsutter** [comment]
But if one is a member then the others must be too, else they are no longer overloads.

**Comment by hsutter** [comment]
Editors call: We have reviewed this but we think the current guidance is good. The workaround you suggest is a workaround that adds another function declaration; people can do that but we don't recommend it.

**Comment by viboes** [comment]
We disagree here also.

Having more reasons to add more function members is not the way I want to go myself. I believe the C++ community should go towards more free functions than member functions. And for me having that one of the overloads needs to have access to the private data, is not enough to add all the possible overloads as members functions,as we have a better option that consists in defining the overload as free functions. 

This could reduce the scope of what each function has access to, and this is what we want to have a better local understanding of a program.  IMO, this guideline is in some way in contradiction with the guideline that ask for reducing the scope of the data/functions, as we are increasing the scope of the private data artificially, by adding functions that don't need to provide access to those private parts.
